helpers/lib_tasks_pytest.py

```diff
@@ -874,6 +874,121 @@ def run_coverage_report(  # type: ignore
         _publish_html_coverage_report_on_s3(aws_profile)
 
 
+def _get_inclusion_settings(target_dir: str) -> Tuple[str, Optional[str]]:
+    """
+    Determine include/omit glob patterns for the coverage report.
+
+    :param target_dir: directory for coverage stats; use "." to indicate all directories
+    :return: tuple (include_in_report, exclude_from_report) where
+        - include_in_report is the glob pattern to include
+        - exclude_from_report is a comma‑separated glob pattern to omit, or ``None``
+    """
+    if target_dir == ".":
+        include_in_report: str = "*"
+        exclude_from_report: Optional[str] = None
+        if hserver.skip_submodules_test():
+            submodule_paths: List[str] = hgit.get_submodule_paths()
+            exclude_from_report = ",".join(
+                f"{path}/*" for path in submodule_paths
+            )
+    else:
+        include_in_report = f"*/{target_dir}/*"
+        exclude_from_report = None
+    return include_in_report, exclude_from_report
+
+
+def _run_coverage(
+    ctx,
+    suite: str,
+    target_dir: str,
+    generate_html_report: bool,
+) -> str:
+    """
+    Internal helper to run coverage for a given suite (fast/slow) and emit XML.
+
+    :param suite: one of "fast" or "slow"
+    :param target_dir: coverage target directory
+    :param generate_html_report: whether to produce HTML output
+    :return: filename of the suite‑specific coverage data file
+    """
+
+    if suite not in ("fast", "slow"):
+        raise ValueError(
+            f"Unsupported suite='{suite}'. Expected 'fast' or 'slow'."
+        )
+    test_cmd = f"invoke run_{suite}_tests --coverage -p {target_dir} -- --junitxml=reports/junit_{suite}.xml -o junit_family=legacy"
+    coverage_file = f".coverage_{suite}_tests"
+    hlitauti.run(ctx, test_cmd, use_system=False)
+    hsystem.system(f"mv .coverage {coverage_file}")
+    hdbg.dassert_file_exists(coverage_file)
+    include_in_report, exclude_from_report = _get_inclusion_settings(target_dir)
+    report_cmd: List[str] = [
+        "coverage erase",
+        f"coverage combine --keep {coverage_file}",
+    ]
+    report_stats_cmd: str = (
+        f"coverage report --include={include_in_report} --sort=Cover"
+    )
+    if exclude_from_report:
+        report_stats_cmd += f" --omit={exclude_from_report}"
+    report_cmd.append(report_stats_cmd)
+    if generate_html_report:
+        report_html_cmd: str = f"coverage html --include={include_in_report}"
+        if exclude_from_report:
+            report_html_cmd += f" --omit={exclude_from_report}"
+        report_cmd.append(report_html_cmd)
+    report_cmd.append("coverage xml -o coverage.xml")
+    full_report_cmd: str = " && ".join(report_cmd)
+    docker_cmd_: str = f"invoke docker_cmd --use-bash --cmd '{full_report_cmd}'"
```

Comment:
No need for these type hints. The linter will infer it automatically. It tells you when it can't